Sequencer: Introduce unit tests + refactoring and live recording

ggVGc opened the pull request:
Reimplementation of https://github.com/datomusic/duo-imxrt/pull/97

This records notes while the sequencer is playing, but the input doesn't feel right to me.
Some tweaking is needed of which step the note is recorded into, and sometimes two notes are recorded because we also add notes which are held when keys are held.

So, some more work is needed to make this feel right, but we can at least evaluate if this is roughly the behaviour we want?

**TODO:**
- [x] Revert changes to cursor behaviour (to be like on `main`)
- [x] The random implementation when stopped is still not completely correct. The cursor jumps to a random step and quickly advances to the step after that.
- [x] Introduce zones. Early, middle, late.
- [x] Try making middle zone dead.
- [x] When sending a MIDI start message the first note is not played (DIN/USB)
- [x] Note doesn't play on advance with RANDOM or BOOST button
- [x] When no clock is received the cursor is always displayed. On DUO 1 the cursor is only displayed when playing a note.
**Re: the cursor isssue:** I think both implementations have their downsides. DUO 1 doesn't show the current step but DUO 2 doesn't show a visual response when you press a key. Either we make DUO 2 behave in the same way or we add a different behaviour as follows: the cursor is always displayed in the form of the current step lighting up in white. Whenever you play a note on the keyboard the current step will light up in the colour of the note. This way we will have the advantages of both implementations.

**Issues when sequencer is stopped**
- [x] Last note priority is gone when playing the keyboard
- [x] Current step indicator is invisible when the play button is lit

**Issues when sequencer is playing:**
- [x] Current step indicator does not light up when a note is disabled
- [x] The arpeggiator sometimes plays an empty note or very high pitched note when holding down many keys and changing keys held
- [x] Notes added to the arpeggiator should be handled more smartly. If a lower note is added than the currently playing note, finish the arpeggio first before adding the new note. If a higher note is added, immediately add the new note to the arpeggio and play it in its normal order

ggVGc commented:
@DavidMenting Please try this out, and read the description for current state.

DavidMenting commented:
This live recording does add a nice touch of responsiveness and playability to the sequencer. 

The double triggering of notes when you play before the step boundary does feel unnatural. I can think of two ways in which we can take control of that issue:

1. Add a dead zone before the next step in which the note is not immediately played but only entered into the sequencer. This is comparable to the existing behaviour but only for a short time period.
2. Always play  a separate timeline that dictates whether a played note ends up in the previous sequencer slot, in the next sequencer slot or nowhere (only play)

There is a new bug here: when pressing the boost key when the sequencer is stopped, a note is played but it remains on. It should turn off when released.

ggVGc commented:
@DavidMenting I didn't have time to finish the implementation of live note recording, but this branch now contains a refactored Sequencer that can be unit tested, and a basic test runner using Unity.
I ran it on my Duo, and things seem to be working fine for me still after the refactoring.

I will have a go tomorrow at doing the proper implementation of note recording.

DavidMenting commented on firmware/seq.cpp:
```diff
@@ -0,0 +1,131 @@
+
+#define INITIAL_VELOCITY 100
+
+Sequencer::Sequencer(Callbacks callbacks) : callbacks(callbacks) {
```
Let's install callbacks in the same way callbacks are added for MIDI handling:
Sequencer.setHandleNoteOn() and Sequencer.setHandleNoteOff()

ggVGc replied:
We can change it, but it'll add some more complexity then. We'll need to check for `null` each time before calling them, and we increase the dynamic behaviour of the sequencer.
In this commit I just made the callbacks references instead, meaning we can statically be sure they're never null: https://github.com/datomusic/duo-imxrt/pull/114/commits/268aee14448ce2bb22319eb14dc5da9e0c3d891c

Unless we have a real use-case for dynamically switching the callbacks during runtime, I don't think we should have the dynamic behaviour. But, if you still prefer the setters, I'll change it.

DavidMenting replied:
there is no need for dynamic switching (yet?) but having an interface consistent with other classes is beneficial.

ggVGc replied:
I didn't do anything in regards to this yet, but I think it's something we can leave until just before finishing this up since it's more of a stylistic concern.

ggVGc commented:
@DavidMenting This is now in a somewhat good state, I believe.
There's still some strange behaviour when inputing a note, where it's played for too long which needs to be fixed for the live recording feature to be complete.

However, the main work here has been:
- Isolate the sequencer functionality in order to make it testable.
- Bring in Unity as a testing framework and write tests around the main functionality of the sequencer.
- Improve the interface of the sequencer and simplify usages of it.
- Rewriting a fair bit of the sequencer internals to achieve the live note recording behaviour.

This should hopefully make it easier to bring the live recording into a complete state, and also be a guideline for future work on the Drum.
There are many things that can be optimized and simplified within the sequencer class, and we might need to mess around with it if this branch produces flicker on other devices than mine. But, with the tests, this work should be much faster since we should be able to trust that we're not breaking functionality by the internal refactoring.
Not everything is covered yet of course, but new tests should be fairly easy to write in cases where we find a regression.

ggVGc commented:
@DavidMenting I think this is now in a state where the sequencer behaves well (at least from my subjective testing).
Instead of writing a new test, I realised the old tests actually had the bad behaviour encoded in them so I adjusted them to be correct and the current implementation passes.
I think the implementation itself is fairly messy, but it's a mix of time pressure and not wanting to change old things too much.

I have also added automatic GitHub actions to run the tests on each push, which you can see by the green checkmarks here at the bottom.

DavidMenting commented:
Are you sure your latest main.cpp is included in the last commit? I get compilation errors

ggVGc commented:
No, the merge went wrong when rebasing on `main` it seems. Fixing it now.

ggVGc commented:
@DavidMenting Should be fixed now, and produces no flicker on my machine

DavidMenting commented:
Issues when sequencer is stopped
- [ ] Last note priority is gone when playing the keyboard
- [ ] Current step indicator is invisible when the play button is lit

Issues when sequencer is playing
- [ ] Current step indicator does not light up when a note is disabled
- [ ] The arpeggiator sometimes plays an empty note or very high pitched note when holding down many keys and changing keys held
- [ ] Notes added to the arpeggiator should be handled more smartly. If a lower note is added than the currently playing note, finish the arpeggio first before adding the new note. If a higher note is added, immediately add the new note to the arpeggio and play it in its normal order

DavidMenting commented:
I've added a (private) forum discussion about this sequencer behaviour https://forum.dato.audio/t/dato-duo-in-sequencer-notes/966

ggVGc commented on platform/brains2/src/duo/main.cpp:
```diff
@@ -286,24 +284,21 @@ static void main_loop(){
-      pots_read();
-      keys_scan(); // 14 or 175us (depending on debounce)
-      #ifdef DEV_MODE
+      if (millis() - frame_time > frame_interval) {
```
The removal of the throttling is what caused the issue with the Env slider LED. I'm not sure exactly why, but I see no problem with adding it again for now (since this is what we have on `main` right now also), so won't dig deeper into it.

DavidMenting replied:
There is actually an interesting explanation to this. The env slider led reads the peak audio value between function calls to determine its brightness. If you read the peak audio volume at a very fast rate you are actually sampling the audio signal, so not only getting the peaks but also the samples between peaks. Limiting the rate at which peak() is called to sub audio rates prevents this. 

There is a tradeoff: faster sampling leads to smoother fades but the brightness will be influenced by low frequency notes.

DavidMenting commented:
Maybe the correct main.cpp is not commited:

```
9: error: 'keyboard_to_note' was not declared in this scope; did you mean 'keyboard_set_note'?
  315 |         keyboard_to_note();
      |         ^~~~~~~~~~~~~~~~
      |         keyboard_set_note
```

ggVGc commented:
@DavidMenting Yeah, it was a result of me merging through the GH interface the other day. Fixed in latest commit, but I'll also take another look at main to see what else might have gotten screwy in the merge.

DavidMenting commented:
In this branch the blinking behaviour of the cursor is different than in main. When playing it works fine but when stopped the next step doesn't automatically light up when advancing the cursor (for example by pressing RANDOM or BOOST or a keyboard key). It's only worth fixing if we go for this new sequencer behaviour though.

ggVGc commented:
> In this branch the blinking behaviour of the cursor is different than in main. When playing it works fine but when stopped the next step doesn't automatically light up when advancing the cursor (for example by pressing RANDOM or BOOST or a keyboard key). It's only worth fixing if we go for this new sequencer behaviour though.

I see two differences in this branch compared to `main` when stopped:
- Holding a note:
  - `main`: Current note lights up until the button is released.
  - This branch: Indicator immediately moves to next step.
- Pressing `random`:
  - `main`: Indicator moves to the next step. Same as pressing `boost`.
  - This branch: Indicator moves to a random step.
  
The first I think is something we could fix to make it consistent with old behaviour.
The second seems more like a bug on `main` to me?

DavidMenting commented:
This works great! With some tweaks to timing I think we can prepare this to become the new behaviour

The refactor of the sequencer and tempohandler does result in weird behaviour with the MIDI clock. The DUO no longer aligns nicely to MIDI clock and the tempo knob does not do what it should. I'll try to split this out into separate issues so we can tackle them one by one.

ggVGc commented:
The latest commit fixes the indicator when moving to the next step by pressing `BOOST`.

ggVGc commented:
The behaviour when holding a note when stopped should also be fixed now.

DavidMenting commented:
- [ ] Note doesn't play on advance with RANDOM or BOOST button
- [ ] When no clock is received the cursor is always displayed. On DUO 1 the cursor is only displayed when playing a note. 

Re: the cursor isssue. I think both implementations have their downsides. DUO 1 doesn't show the current step but DUO 2 doesn't show a visual response when you press a key. Either we make DUO 2 behave in the same way or we add a different behaviour as follows: the cursor is always displayed in the form of the current step lighting up in white. Whenever you play a note on the keyboard the current step will light up in the colour of the note. This way we will have the advantages of both implementations.

ggVGc commented:
Merged the tempo-fix PR into this, and moved new issues named here and in the tempo-fix branch into the description of this PR.

ggVGc commented:
@DavidMenting  The last commit implements the changes to the cursor, but I think it's a bit weird for a few reasons:
- The situation we're dealing with is when there is no clock received while the sequencer is running, after having received MIDI clock, and a user plays the keyboard. This doesn't seem like a very common situation.
- The current step now does not show the step gate length (since the current step is always white, unless a keyboard note is played)
- When a keyboard note is played, the colour of the step shows the pressed note, but depending on which tick we are on, it will be recorded in this step, next step, or not at all. This is fairly confusing and would seem like a bug.
- These changes seem to impact normal playing while only improving a situation which I think is an edge-case.
- It feels to me like we're now looking at "no clock being received" as "stopped", which is not the case. If we want to handle the situation where a clock hasn't been received for a while, I think we need to introduce some kind of paused state.

ggVGc commented:
@DavidMenting The previous commits addresses the other issues, and I ticked the boxes of what should be fixed in the description.

DavidMenting commented:
Alright, we're getting there. The random implementation when stopped is still not completely correct. The cursor jumps to a random step and quickly advances to the step after that.

ggVGc commented:
@DavidMenting The behaviour random behaviour should now be fixed, and the cursor behaviour should be what we have in `main`.

DavidMenting commented:
Yes! Random and the cursor are behaving fine. Let's try to get the FlexIO leds into this branch to get rid of the flicker and we should  be good to go

ggVGc commented:
@DavidMenting This is based on latest `main` now, and I think if everything works as it should, we can merge it.
